ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java

```diff
@@ -186,6 +187,60 @@ public NetworkClient(Selector selector, NetworkConfig networkConfig, NetworkMetr
     return sends;
   }
 
+  /**
+   * Warm up connections to dataNodes in a specify time window.
+   * If a connection established successfully, it's a successConnection.
+   * If a connection failed(for example, target dataNode offline), it's a failedConnection.
+   * If a connection breaks after successfully established, it's counted as a failedConnection. This impact the
+   * counting in the method, but this is tolerable as other good/bad connections can be handled in next selector.poll().
+   * If a connection established after this time window of timeForWarmUp, it can be handled in next selector.poll().
+   * @param dataNodeIds warm up target nodes.
+   * @param connectionCountPerDataNode number of connections would like to establish in the warmup.
+   * @param timeForWarmUp max time to wait for connections' establish.
+   * @return number of connections established successfully.
+   */
+
+  public int warmUpConnections(List<DataNodeId> dataNodeIds, int connectionCountPerDataNode, long timeForWarmUp) {
+    int expectedConnections = 0;
+    logger.info("Connections warm up start.");
+    for (DataNodeId dataNodeId : dataNodeIds) {
+      // only warm up connections to dataNodes in local DC.
```

Comment:
this comment seems out of place here since the caller provides the `DataNodeId`s that need to be warmed up